File: frontend/src/types/agent.ts
export type ToolConfig = {
  id: number
  slug: string
  name: string
  description: string
  parameters_schema: string
  handler_type: string
  handler_config: string
  is_enabled: boolean
  created_at: string
}

export type SkillConfig = {
  id: number
  slug: string
  name: string
  description: string
  prompt_text: string
  is_enabled: boolean
  created_at: string
}

export type AgentConfig = {
  id: number
  slug: string
  name: string
  description: string
  icon: string
  system_prompt: string
  user_prompt_template: string
  model: string
  temperature: number
  max_tokens: number
  max_iterations: number
  is_enabled: boolean
  sort_order: number
  created_at: string
  updated_at: string
  tools: ToolConfig[]
  skills: SkillConfig[]
}

export type AgentCreate = {
  slug: string
  name: string
  description?: string
  icon?: string
  system_prompt?: string
  user_prompt_template?: string
  model?: string
  temperature?: number
  max_tokens?: number
  max_iterations?: number
  is_enabled?: boolean
  sort_order?: number
  tool_ids?: number[]
  skill_ids?: number[]
}

export type AgentUpdate = Partial<AgentCreate>

export type ToolCallTrace = {
  name: string
  arguments: Record<string, unknown>
  result: string
  duration_ms: number
}

export type TokenUsage = {
  prompt_tokens: number
  completion_tokens: number
  total_tokens: number
}

export type AgentRunRequest = {
  context: Record<string, unknown>
  history?: Record<string, unknown>[]
}

export type AgentRunResult = {
  content: string
  tool_calls_trace: ToolCallTrace[]
  token_usage: TokenUsage
  iterations: number
  duration_ms: number
}

export type UsageSummary = {
  agent_slug: string
  model: string
  total_requests: number
  total_prompt_tokens: number
  total_completion_tokens: number
  total_tokens: number
  total_tool_calls: number
  avg_duration_ms: number
}

export type ChatSession = {
  id: number
  agent_id: number
  title: string
  created_at: string
  updated_at: string
}

export type ChatSessionCreate = {
  agent_id: number
  title?: string
}

export type ChatSessionUpdate = {
  title: string
}

export type ChatMessage = {
  id: number
  session_id: number
  role: string
  content: string
  tool_calls: string | null
  tool_results: string | null
  token_usage: string | null
  created_at: string
}

export type ChatRequest = {
  query: string
  context?: Record<string, unknown>
}

export type SSEEventType = 'meta' | 'thinking' | 'tool_call' | 'tool_result' | 'content' | 'status' | 'done' | 'error'

export type SSEEvent = {
  type: SSEEventType
  data: unknown
}

File: backend/src/schemas/agent.py
from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field


class AgentCreate(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    slug: str = Field(min_length=1, max_length=64)
    name: str = Field(min_length=1, max_length=128)
    description: str = Field(default='')
    icon: str = Field(default='bot', max_length=32)
    system_prompt: str = Field(default='')
    user_prompt_template: str = Field(default='')
    model: str = Field(default='gpt-4.1-mini', max_length=128)
    temperature: float = Field(default=0.2, ge=0.0, le=2.0)
    max_tokens: int = Field(default=2048, ge=1, le=32768)
    max_iterations: int = Field(default=10, ge=1, le=50)
    is_enabled: bool = Field(default=True)
    sort_order: int = Field(default=0, ge=0)
    tool_ids: list[int] = Field(default_factory=list)
    skill_ids: list[int] = Field(default_factory=list)


class AgentUpdate(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    slug: str | None = Field(default=None, min_length=1, max_length=64)
    name: str | None = Field(default=None, min_length=1, max_length=128)
    description: str | None = None
    icon: str | None = Field(default=None, max_length=32)
    system_prompt: str | None = None
    user_prompt_template: str | None = None
    model: str | None = Field(default=None, max_length=128)
    temperature: float | None = Field(default=None, ge=0.0, le=2.0)
    max_tokens: int | None = Field(default=None, ge=1, le=32768)
    max_iterations: int | None = Field(default=None, ge=1, le=50)
    is_enabled: bool | None = None
    sort_order: int | None = Field(default=None, ge=0)
    tool_ids: list[int] | None = None
    skill_ids: list[int] | None = None


class AgentConfig(BaseModel):
    id: int
    slug: str
    name: str
    description: str
    icon: str
    system_prompt: str
    user_prompt_template: str
    model: str
    temperature: float
    max_tokens: int
    max_iterations: int
    is_enabled: bool
    sort_order: int
    created_at: str
    updated_at: str
    tools: list = Field(default_factory=list)
    skills: list = Field(default_factory=list)
